src/actions/pipelines.js

```diff
@@ -114,19 +116,21 @@ export function loadInitialPipelineData() {
  */
 export function loadPipelineData(pipelineID) {
   return async function (dispatch, getState) {
-    const { dataSource, pipeline } = getState();
-    if (pipelineID && pipelineID === pipeline.active) {
+    const state = getState();
+    if (pipelineID && pipelineID === state.pipeline.active) {
       return;
     }
-    if (dataSource === 'json') {
+    if (state.dataSource === 'json') {
       dispatch(toggleLoading(true));
       // Remove the previous graph to show that a new pipeline is being loaded
       dispatch(toggleGraph(false));
       const url = getPipelineUrl({
-        main: pipeline.main,
+        main: state.pipeline.main,
         active: pipelineID,
       });
-      const newState = await loadJsonData(url).then(preparePipelineState);
+      const newState = await loadJsonData(url).then((data) =>
+        normalizeData(state, data)
```

Comment:
Could there be any problems here when loading is slow and the `state` becomes stale? Should it call `getState()` again?